util/topgen/merge.py

```diff
@@ -296,6 +335,140 @@ def amend_xbar(top, xbar):
         xbar_adddevice(top, topxbar, device)
 
 
+def xbar_cross(xbar, xbars):
+    """Check if cyclic dependency among xbars
+
+    And gather the address range for device port (to another Xbar)
+
+    @param node_name if not "", the function only search downstream
+                     devices starting from the node_name
+    @param visited   The nodes it visited to reach this port. If any
+                     downstream port from node_name in visited, it means
+                     circular path exists. It should be fatal error.
+    """
+    # Step 1: Visit devices (gather the address range)
+    log.info("Processing circular path check for {}".format(xbar["name"]))
+    addr = []
+    for node in [
+            x for x in xbar["nodes"]
+            if x["type"] == "device" and "xbar" in x and x["xbar"] == False
+    ]:
+        # TODO: Can this be simplified? (Merging contiguous into one)
+        addr.append((node["base_addr"], node["size_byte"]))
+
+    # Step 2: visit xbar device ports
+    xbar_nodes = [
+        x for x in xbar["nodes"]
+        if x["type"] == "device" and "xbar" in x and x["xbar"] == True
+    ]
+
+    # Now call function to get the device range
+    for node in xbar_nodes:
+        xbar_addr = xbar_cross_node(node["name"], xbar, xbars, visited=[])
```

Comment:
@eunchan i'm not sure if i'm reading this right...but from the above, isn't xbar_nodes a list of the actual device xbars, and xbar the current one we are looking at?  So for example, on the first iteration, `node["name"]` would be `peri`, and xbar the main xbar? 

I got confused mainly because of the naming in `xbar_cross_node`.  It looks we are searching for whatever matches the peri name (in this case), and calling that host_xbar? while main becomes device?

Reply:
That's true, the `node["name"]` is used to find the host_xbar and its connection. The assumption here is that there's only one connection from crossbar A to crossbar B. 

To clarify, `device_xbar` is the crossbar has a device port with name as `node["name"]`. `host_xbar` is the crossbar has a host port with name as `node["name"]`.

Reply:
ahhhh.
could i ask you add a comment to clarify that? That definitely makes things much clearer.

Reply:
OK. I will

Reply:
Revised